infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,316 @@
+import lombok.EqualsAndHashCode.Exclude;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.common.result.Result;
+import org.corfudb.infrastructure.log.StreamLog;
+import org.corfudb.infrastructure.log.statetransfer.batch.TransferBatchRequest;
+import org.corfudb.infrastructure.log.statetransfer.batch.TransferBatchResponse;
+import org.corfudb.infrastructure.log.statetransfer.batch.TransferBatchResponse.FailureStatus;
+import org.corfudb.infrastructure.log.statetransfer.batchprocessor.StateTransferBatchProcessor;
+import org.corfudb.infrastructure.log.statetransfer.streamprocessor.TransferSegmentFailure;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.SegmentState.NOT_TRANSFERRED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.SegmentState.TRANSFERRED;
+
+/**
+ * A class responsible for managing a state transfer on the current node.
+ * It executes the state transfer for each non-transferred segment synchronously.
+ */
+@Slf4j
+@AllArgsConstructor
+public class StateTransferManager {
+
+    /**
+     * States of the segment:
+     * - {@link SegmentState#NOT_TRANSFERRED}: Segment is not transferred.
+     * - {@link SegmentState#TRANSFERRED}: Segment was transferred fully.
+     * - {@link SegmentState#RESTORED}: Segment was restored, and is present in the current layout.
+     * - {@link SegmentState#FAILED}: The state transfer of a segment has failed.
+     */
+    public enum SegmentState {
+        NOT_TRANSFERRED,
+        TRANSFERRED,
+        RESTORED,
+        FAILED
+    }
+
+    /**
+     * A data class that represent one segment to be transferred.
+     */
+    @EqualsAndHashCode
+    @Getter
+    @ToString
+    @Builder
+    @AllArgsConstructor(access = AccessLevel.PRIVATE)
+    public static class TransferSegment implements Comparable<TransferSegment> {
+        /**
+         * Start address of a segment.
+         */
+        private final long startAddress;
+        /**
+         * End address of a segment.
+         */
+        private final long endAddress;
+        /**
+         * A future that holds the status of a transfer of a segment.
+         */
+        private final TransferSegmentStatus status;
+
+        @Override
+        public int compareTo(TransferSegment other) {
+            return Long.compare(this.getStartAddress(), other.getStartAddress());
+        }
+
+
+        /**
+         * Whether a current segment overlaps with another segment.
+         *
+         * @param other Other segment.
+         * @return True, if overlap.
+         */
+        public boolean overlapsWith(TransferSegment other) {
+            return other.getStartAddress() <= this.getEndAddress();
```

Comment:
The transfer segment can never be the last segment since its infinite. It's filtered out when the transfer segments are created.
However, there was a bug (in split segments code) that was breaking the abstraction of a transfer segment, resulting in the end address being -1, which would require this check. I fixed this bug now and the abstraction is now more concrete, hence this check is not needed. I will address what was happening in the comments to the PR once I'll push the latest changes.